### ISSUE TITLE
Crash on screen rotation with SelectionController.

### ISSUE THREAD

**Issue by aveuiller** [dkharrat/NexusDialog#11]
The application crash on a screen rotation When a SelectionController is in the currently shown Form with an EditTextController or a SearchableSelectionController, with the Error:

    com.github.dkharrat.nexusdialog.sample E/AndroidRuntime: FATAL EXCEPTION: main
        Process: com.github.dkharrat.nexusdialog.sample, PID: 7355
        java.lang.RuntimeException: Unable to start activity ComponentInfo{com.github.dkharrat.nexusdialog.sample/com.github.dkharrat.nexusdialog.sample.SimpleExample}: java.lang.ClassCastException: android.widget.TextView$SavedState cannot be cast to android.widget.Spinner$SavedState
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2184)
            [...]
        Caused by: java.lang.ClassCastException: android.widget.TextView$SavedState cannot be cast to android.widget.Spinner$SavedState
            at android.widget.Spinner.onRestoreInstanceState(Spinner.java:762)
            [...]

The value of `SPINNER_ID` and `EDIT_TEXT_ID` seems to be mixed-up on the redraw (I don't really see why though), so changing the value of `SPINNER_ID` will solve the problem.

**Comment by dkharrat** [comment]
Good catch! Seems like the reason is that IDs are being reused across controllers, which apparently affects restoring state of views for `onSaveInstanceState` and `onRestoreInstanceState`. That seems to contradict the Android documentation, which states IDs don't have to be unique, which is strange. I see 3 possible ways to fix this:

1. Change the IDs manually to be unique for each controller. The issue with this approach is that it's difficult to completely guarantee unique IDs, as whatever ID we choose can technically conflict with an ID generated by Android for layouts. Also, apps implementing a custom controller need to ensure they use a custom ID, which could be error-prone.
2. Use [`View#generateViewId`](http://developer.android.com/reference/android/view/View.html#generateViewId%28%29) to solve the issue in 1. However, this requires API 17+, which will be an issue for NexusDialog. One way to address that is to copy the unique ID generation code into NexusDialog so we're not dependent on the API level for this functionality.
3. Create the field views via an XML layout rather than programmatically, in order to rely on the Android aapt tool for ensuring ID uniqueness.

I prefer option 3, since it also provides the advantage of allowing an app to override the define view layout if needed. However, we should go with option 2 for dynamic elements, such as the checkboxes.

What are your thoughts?

**Comment by aveuiller** [comment]
In fact, IDs doens't have to be unique between different layouts, the problem is that when rebuilding the form view, android seems to use the global layout of the Activity and returns the first view with the given ID.

Using xml defined layout for static fields like textView might be a good idea.

On the other hand, the method generateViewId() cannot be used as-is (it uses a private attribute of View to check the availability of the next ID), but we can use the same idea by returning the first ID for which `((Activity) context).findViewById(id)` returns `null` on the FormElementController.

**Comment by dkharrat** [comment]
The main issues I see with using `findViewById`:

1. It's not guaranteed that the user passes in an `Activity` instance for the `context`. They could technically pass in the Application context.
2. Although probably not an issue in the most of the cases, there might be a performance penalty when finding views, especially if there are a lot of nested views.

As for using the same implementation as `View#generateViewId()`, I think we can also copy the same attribute to NexusDialog. In other words, NexusDialog maintains its own `AtomicInteger` to store the last generated id. This approach will require that users needing a unique ID for their own views use the NexusDialog method to generate a unique ID for the Form activity, but this can be documented. In any case, how we generate the unique ID is an implementation detail and shouldn't affect the overall solution.

So to summarize the proposed solution:

1. Create field views via an XML layout rather than programmatically (to serve two purposes: allow clients to override views if needed, and also have Android generate the id for us).
2. For views that need to be instantiated programmatically, use a method provided by NexusDialog to provide a unique ID (unique within the Form activity). I think this method should live in `FormController`.

**Comment by aveuiller** [comment]
There is only one thing that bother me for this solution:
For what I understand from the usage of `View#generateViewId()`, every generated ID goes through this method, enabling Android to know what the next ID needs to be. 

With the proposed solution (on devices under API 17), most of the IDs will be generated through xml layouts (hence Android), and the others (the dynamic fields) through the FormController method.

The problem here is that the two `AtomicIntegers` won't be linked, possibly creating the same ID for two elements.

Having said that, depending on the implementation, the probability of having the same ID for two elements on the user's FormActivity may be low enough to use this solution.

**Comment by dkharrat** [comment]
My understanding is that IDs generated for the xml layouts will have the higher 8-bits set (i.e. 0xff000000). You can see this by checking the `R.java` file. This is why the `View#generateViewId()` has this check:

```
if (newValue > 0x00FFFFFF) newValue = 1;
```

Basically, it ensures the higher 8-bits are always 0. So, as long as we have the same logic, IDs generated from the xml layouts will not collide with the IDs generated by this method.

**Comment by aveuiller** [comment]
Ok, so the ID generated at compile time are completely separated from the runtime generated ones. On this case, everything is good for me, thanks for the explanation. :)

Will you implement this fix, or should I?

**Comment by dkharrat** [comment]
Please go ahead and implement it. Thanks!

**Comment by brettalcox** [comment]
I'm still having this issue actually. Is this fix included in the build linked with Gradle (0.3.0)? Same stack trace and everything.

**Comment by dkharrat** [comment]
It's fixed in master, but hasn't been released yet. There's been quite a few changes since the 0.3.0 release. I'll be releasing 0.4.0 tonight, which has the fix.

**Comment by brettalcox** [comment]
Oh alright awesome.

**Comment by dkharrat** [comment]
FYI, 0.4.0 has been released now.
